### PR TITLE
Added support for collections in :class property

### PR DESCRIPTION
**Pull request by zoldar**
That's my first stab at adding support for passing a collection for class property, as mentioned in #138. Is that the right place to implement it?

### CODE REVIEW

**Comment by whodidthis** [comment]
I think it could be useful to also add support for conditional classes, kind of like what react devs had in mind for inline styles https://github.com/reactjs/react-future/blob/master/04%20-%20Layout/04%20-%20Inline%20Styles.md#using-styles.

So remove falsey values to enable `["my-input" (and @active "active")]`.

Also wouldn't mind to have those conditional inline styles implemented heh

**Comment by zoldar** [comment]
@whodidthis Good idea. Added filtering step.

File: src/reagent/impl/template.cljs [hunk 1]
```diff
@@ -73,13 +73,24 @@
                            class))))
     p))
 
+(defn stringify-class [{:keys [class] :as props}]
+  (if (coll? class)
+    (->> class
+         (filter identity)
```
Comment (JulianLeviston) [inline src/reagent/impl/template.cljs]: Why filter identity here? isn't this a no-op, or at best the equiv of seq?

Comment (JulianLeviston) [inline src/reagent/impl/template.cljs]: oh, ok. Nevermind. I understand now - I read the comments.

**Comment by JulianLeviston** [comment]
You can already do this with clojure's built-in functionality. This might be best suited by writing a helper method.

Adding this code adds at best an additional if statement around every single class in our apps. What if we don't want this in our code? At worst, the filter adds quite a bit of code to these collections, and ends in some magic. At best, another conditional.

The more magic we have in our code, the more bloated and less understandable it becomes. One would think this comment isn't necessarily relevant for this patch, but it's actually the little things that matter.

Clojure, and Reagent are awesome precisely because they are tiny, comprehensive and understandable. Precisely because of the magic they *don't* have.

**Comment by zoldar** [comment]
@JulianLeviston I agree that adding feature increasing code complexity just because it may prove useful is not a good idea. But in this actual case, maybe the added complexity is balanced out by how often it's found useful? It would be probably best if others had their say on this before deciding what to do about the PR.

**Comment by JulianLeviston** [comment]
@zoldar Sure. It's only my two cents, afterall :)

**Comment by Deraen** [comment]
`(filter identity)` is unnecessary as nil is anyway outputted as empty string, it might add some unnecessary spaces but that shouldn't be a problem.
Also instead of `interpose` and `apply str`, `clojure.string/join` would work.

```clj
(defn stringify-class [{:keys [class] :as props}]
  (if (coll? class)
    (assoc props :class (string/join " " class))
    props))
```

**Comment by zoldar** [comment]
@Deraen Good points. Also, `string/join` actually does omit putting separator after nil values - neat.

**Comment by whodidthis** [comment]
`false` is printed as `"false"` though, or is there some other more obvious way than
`(and @my-atom "my-class")`

**Comment by Deraen** [comment]
`(if @my-atom "my-class")` but I guess it would be best if `and` worked also.

**Comment by whodidthis** [comment]
Right, `when` and `if`. I guess i translated `&&` a bit too directly heh.

Although i guess i would still go with filtering out falses, `(and false 1)` returns `false` and may trick people although maybe not too many use a class called false.

Also using `and` would still make sense in `(and @dirty @warning "warning")` etc in my opinion

**Comment by zoldar** [comment]
@whodidthis @Deraen I also think that it will be best and least surprising if all falsey values are removed. Brought back the filtering by identity step.

**Comment by mike-thompson-day8** [comment]
I'm in favour of merging this.

**Comment by holmsand** [comment]
Yes, I agree: it might be convenient to have. Let's do this after 0.5.1. (There should probably be some tests for this, though...)

@mike-thompson-day8 Yes, we should definitely have some visible documentation of especially syntax. Ideally that should be on the website – as that is most visible, I guess?

**Comment by zoldar** [comment]
Added tests for the feature.

**Comment by zoldar** [comment]
Finally got my s**t together and learned how to use Git properly (hopefully). Now everything is squahsed into a single, nice commit.

**Comment by runningskull** [comment]
:+1:

**Comment by bsteuber** [comment]
+1

**Comment by zefirka** [comment]
+1

File: src/reagent/impl/template.cljs [hunk 2]
```diff
@@ -73,13 +73,23 @@
                            class))))
     p))
 
+(defn stringify-class [{:keys [class] :as props}]
+  (if (coll? class)
```
Comment (danielcompton) [inline src/reagent/impl/template.cljs]: Should this be sequential? instead of coll? maps and sets are also coll?

Comment (zoldar) [inline src/reagent/impl/template.cljs]: I have used `coll?` in order to support sets actually (which is not covered by tests as I see it now, bummer). Do you think it's impractical/unnecessary?

Comment (bsteuber) [inline src/reagent/impl/template.cljs]: Supporting sets is good, so just leave it that way.
I guess the remark meant to ask how to deal with maps. Maybe throwing an
error would be best, but probably "undefined behavior" is okay too - I
mean, who would do that anyways :-P

Adrian Gruntkowski <notifications@github.com> schrieb am Mi., 28. Okt. 2015
10:19:

**Comment by mike-thompson-day8** [comment]
@holmsand  are you okay with this being merged?   It has tests but no docs.

**Comment by danielcompton** [comment]
It would be good to add tests for sets if that usage is going to be supported.

**Comment by zoldar** [comment]
Added test for sets and a short mention of the feature in `README.md`

**Comment by zoldar** [comment]
Oops, hit some conflicts. Will try to sort it out later.

**Comment by holmsand** [comment]
@zoldar Sorry about that – I've just merged a whole lot of code all over the place :)

I think this patch is a good idea btw, so we should merge it.

**Comment by Deraen** [comment]
Maps where mentioned in some comment. I can think a way how maps could be useful for setting classes:

```clj
;; active? true, odd? false
:class [(if active? "active") (if odd? "odd")]
:class {:active active? :odd odd?}
;; => "active"

;; active? true, odd? true
:class [(if active? "active") (if odd? "odd")]
:class {:active active? :odd odd?}
;; => "active odd"
```

**Comment by bsteuber** [comment]
I like that way of using maps - it's like Angular's ng-class directive and
would in indeed be useful for conditionals.

**Comment by mike-thompson-day8** [comment]
@Deraen I'm not a fan of the map version.  Too much magic, for not enough benefit.  

By magic I mean:  an experienced clojurescript developer is not going to immediately understand what's going on without some coaching.  There'd have to be a HUGE upside to warrant the overhead of puzzling people.  Remember this is a relatively minor, non essential, convenience feature.

**Comment by zoldar** [comment]
I'm with @mike-thompson-day8 on the map suggestion. That's a bit too much magic for me too.

P.S.: Resolved the conflicts.

**Comment by mike-thompson-day8** [comment]
@holmsand  are you still okay with this being merged?

**Comment by holmsand** [comment]
In principle yes. But first I'd like to make sure that it doesn't have any adverse effects on performance. I think it can be made quite fast, but I'd like to (finally) get a 0.6.0-alpha out the door first, this week I sincerely hope...

**Comment by ericnormand** [comment]
Ping! Any idea when this will be released?

**Comment by aerosol** [comment]
@holmsand ping

File: README.md [hunk 3]
```diff
@@ -55,6 +55,12 @@ can be written as:
 [:div>p>b "Nested Element"]
 ```        
 
+The `:class` attribute accepts collections in addition to plain string.
```
Comment (GetContented) [inline README.md]: Suggestion: improve language.

"The :class attribute can accept either a collection or a string"

**Comment by zoldar** [comment]
Applied feedback and rebased.

**Comment by Deraen** [comment]
Looks good. I was waiting for #313 before merging some PRs, as I broke CI config at some point, but as this has good tests, I'm sure this will work after I merge new test runner.

**Comment by pesterhazy** [comment]
Fantastic to see this land - a much requested feature!

**Comment by MichaelFHScott** [comment]
I can't seem to make this work - it separates the classes with commas not spaces so they aren't recognised. Any idea what I'm doing wrong?

**Comment by ericnormand** [comment]
@MichaelFHScott Can you share a quick snippet of code so we can see what you're doing?

**Comment by MichaelFHScott** [comment]
Thanks for the very quick response. It was my stupid mistake - I had accidentally reverted to Reagent 0.7.0.
I'm just discovering Clojurescript and Reagent. It's brilliant!

**Pull request merged**